onnxruntime/core/session/inference_session.h

```diff
@@ -3,130 +3,44 @@
+#include "core/framework/path_lib.h"
+
+#ifdef USE_EIGEN_THREADPOOL
+#include <unsupported/Eigen/CXX11/ThreadPool>
+#endif
 
 namespace ONNX_NAMESPACE {
 class ModelProto;
-}  // namespace ONNX_NAMESPACE
-
-struct OrtCustomOpDomain {
-  std::string domain_;
-  std::vector<OrtCustomOp*> custom_ops_;
-};
+}
 
 namespace onnxruntime {
-class IExecutionProvider;  // forward decl
+class IExecutionProvider;
+class Graph;
+class GraphTransformer;
+class Model;
 class IOBinding;
+class Notification;
+class IExecutor;
+class TaskThreadPool;
 
-class CustomRegistry;
-
-namespace logging {
-class LoggingManager;
-}
-
-/**
-  * Configuration information for a session.
-  */
-struct SessionOptions {
-  //int num_threads; // not used now until we re-introduce threadpools for async execution
-  bool enable_sequential_execution = true;  // TODO: should we default to sequential execution?
-
-  // enable profiling for this session.
-  bool enable_profiling = false;
-
-  // enable the memory arena on CPU
-  // Arena may pre-allocate memory for future usage.
-  // set this option to false if you don't want it.
-  bool enable_cpu_mem_arena = true;
-
-  // the prefix of the profile file. The current time will be appended to the file name.
-  std::basic_string<ORTCHAR_T> profile_file_prefix = ORT_TSTR("onnxruntime_profile_");
-
-  std::string session_logid;                 ///< logger id to use for session output
-  unsigned session_log_verbosity_level = 0;  ///< applies to session load, initialization, etc
-
-  unsigned max_num_graph_transformation_steps = 5;  // TODO choose a good default here?
-
-  // set graph optimization level
-  // allowed levels are 0, 1, 2
-  // 0 -> disable all optimizations
-  // 1 -> enable basic optimizations
-  // 2 -> enable all optimizations
-  unsigned graph_optimization_level = 0;
-
-  // How many threads in the session thread pool.
-  int session_thread_pool_size = 0;
-};
-
-/**
-  * Pre-defined and custom metadata about the model.
-  */
-struct ModelMetadata {
-  std::string producer_name;
-  std::string graph_name;
-  std::string domain;
-  std::string description;
-  int64_t version;
-  std::unordered_map<std::string, std::string> custom_metadata_map;
-};
-
-/**
-  * @brief This is the main class used to Run a model.
-  * Sample simple usage:
-  *  CPUExecutionProviderInfo epi;
-  *  ProviderOption po{"CPUExecutionProvider", epi};
-  *  SessionOptions so(vector<ProviderOption>{po});
-  *  InferenceSession session_object{so};
-  *  common::Status status = session_object.Load(MODEL_URI);
-  *  common::Status status = session_object.Initialize();
-  *
-  *  NameMLValMap feeds;
-  *  feeds.insert({});
-  *  ...
-  *  std::vector<std::string> output_names;
-  *  output_names.insert(...);
-  *  ...
-  *  std::vector<MLValue> fetches;
-  *  common::Status status = session_object.Run(run_options, feeds, output_names, &fetches);
-  *  process the output here...
-  */
-
-class InferenceSession {
+class InferenceSession : public Session {
```

Comment:
quick question: TrainingSession will be a subclass of InferenceSession?

Reply:
Yes, it will.

Reply:
For Training, there could be a separate TrainableSession interface if needed, while the actual implementation TrainingSession is a subclass of InferenceSession.